bee-network/bee-autopeering/src/peering/neighbor.rs

```diff
@@ -0,0 +1,347 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    hash,
+    local::{salt::Salt, Local},
+    peer::{peer_id::PeerId, Peer},
+};
+
+use prost::bytes::{Buf, Bytes};
+
+use std::{
+    cmp, fmt,
+    sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard},
+    vec,
+};
+
+/// The distance between the local entity and a neighbor.
+pub type Distance = u32;
+
+pub(crate) const SIZE_INBOUND: usize = 4;
+pub(crate) const SIZE_OUTBOUND: usize = 4;
+
+/// Decides whether a peer is a suitable neighbor, or not.
+pub trait NeighborValidator
+where
+    Self: Send + Sync + Clone,
+{
+    /// Returns `true` if the given [`Peer`](crate::peer::Peer) is a valid neighbor.
+    fn is_valid(&self, peer: &Peer) -> bool;
+}
+
+// A neighbor is a peer with a distance) metric.
+#[derive(Clone, Debug)]
+pub(crate) struct Neighbor {
+    peer: Peer,
+    distance: Distance,
+}
+
+impl Neighbor {
+    pub(crate) fn new(peer: Peer, distance: Distance) -> Self {
+        Self { peer, distance }
+    }
+
+    pub(crate) fn peer(&self) -> &Peer {
+        &self.peer
+    }
+
+    pub(crate) fn distance(&self) -> Distance {
+        self.distance
+    }
+}
+
+impl fmt::Display for Neighbor {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.peer().peer_id(), self.distance())
+    }
+}
+
+impl Eq for Neighbor {}
+impl PartialEq for Neighbor {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+impl PartialOrd for Neighbor {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Ord for Neighbor {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.distance.cmp(&other.distance)
+    }
+}
+
+impl AsRef<Peer> for Neighbor {
+    fn as_ref(&self) -> &Peer {
+        &self.peer
+    }
+}
+
+#[derive(Clone, Default)]
+pub(crate) struct Neighborhood<const N: usize, const INBOUND: bool> {
+    inner: Arc<RwLock<NeighborhoodInner<N, INBOUND>>>,
+}
+
+impl<const N: usize, const INBOUND: bool> Neighborhood<N, INBOUND> {
+    /// Creates a new empty neighborhood.
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    /// Inserts a peer to the neighborhood.
+    pub(crate) fn insert_neighbor(&self, peer: Peer, local: &Local) -> bool {
+        self.write().insert_neighbor(peer, local)
+    }
+
+    /// Removes a peer from the neighborhood.
+    pub(crate) fn remove_neighbor(&self, peer_id: &PeerId) -> Option<Peer> {
+        self.write().remove_neighbor(peer_id)
+    }
+
+    /// Checks whether the candidate is a suitable neighbor.
+    pub(crate) fn is_preferred(&self, candidate: &Neighbor) -> bool {
+        self.write().is_preferred(candidate)
+    }
+
+    /// Picks the first candidate, that is closer than the currently furthest neighbor.
```

Comment:
These commas are so German 😄

Reply:
I am fine with removing the comma in this case.

Reply:
Removed